### PR TITLE
add sentiment-analysis-cnn

### PR DESCRIPTION
**Pull request by xiaotinghe**

### CODE REVIEW

**Comment by astonzhang** [comment]
Thanks. First, can you make sure that the code can run? I added a few comments in your code.

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+            setattr(self, f'conv_{i}', conv)
+            setattr(self, f'pool_{i}', pool)
+        self.dropout = nn.Dropout(0.5)
+        self.decoder = nn.Dense(num_outputs)
+    def forward(self, inputs):
+        embeddings_static = self.embedding_static(inputs).transpose((1,2,0))
+        embeddings_non_static = self.embedding_non_static(inputs).transpose((1,2,0))
+        embeddings = nd.concat(embeddings_static,embedding_non_static,dim=1)
+        encoding = [
+            nd.flatten(self.get_pool(i)(self.get_conv(i)(x)))
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: What is x?

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            conv = nn.Conv1D(self.FILTER_NUM[i], kernel_size=self.FILTERS[i], strides=1, activation='relu')
+            pool = nn.GlobalMaxPool1D()
+            setattr(self, f'conv_{i}', conv)
+            setattr(self, f'pool_{i}', pool)
+        self.dropout = nn.Dropout(0.5)
+        self.decoder = nn.Dense(num_outputs)
+    def forward(self, inputs):
+        embeddings_static = self.embedding_static(inputs).transpose((1,2,0))
+        embeddings_non_static = self.embedding_non_static(inputs).transpose((1,2,0))
+        embeddings = nd.concat(embeddings_static,embedding_non_static,dim=1)
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: Is it `embedding_non_static` or `embeddings_non_static`?

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+num_outputs = 2
+lr = 0.1
+num_epochs = 1
+batch_size = 10
+embed_size = 100
+num_hiddens = 100
+num_layers = 2
+bidirectional = True
+    
+net = TextCNN(vocab, embed_size, num_hiddens)
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: What is `num_hiddens`? What is `bidirectional`?

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
+# 文本分类：情感分析
+
+文本分类即把一段不定长的文本序列变换为类别。在文本分类问题中，情感分析是一项重要的自然语言处理任务。例如，Netflix或者IMDb可以对每部电影的评论进行情感分类，从而帮助各个平台改进产品，提升用户体验。
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: This paragraph is the same as the previous one. Can you just give a brief intro and highlight the difference?

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+# 文本分类：情感分析
+
+在之前的章节中介绍了卷积神经网络用于计算机视觉领域。
+在本节将介绍如何将卷积神经网络应用于自然语言处理领域。以及参考textCNN模型使用Gluon创建一个卷积神经网络用于文本情感分类。
+
+## 模型设计
+
+卷积神经网络用于自然语言处理，可以将理解为把一个文本用二维图像的方式来表达，每一行是一个词向量，将每一个词纵向排列。
+
+![](../img/embedding.svg)
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 这张图可以不放，因为下面图里已经有了这个内容

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        self.embedding_non_static = nn.Embedding(len(vocab), embedding_size)
+        for i in range(len(self.FILTERS)):
+            conv = nn.Conv1D(self.FILTER_NUM[i], kernel_size=self.FILTERS[i], strides=1, activation='relu')
+            pool = nn.GlobalMaxPool1D()
+            setattr(self, f'conv_{i}', conv)
+            setattr(self, f'pool_{i}', pool)
+        self.dropout = nn.Dropout(0.5)
+        self.decoder = nn.Dense(num_outputs)
+    def forward(self, inputs):
+        embeddings_static = self.embedding_static(inputs).transpose((1,2,0))
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 注释一下为什么这里要transpose，以及transpose后每一维的意义是什么

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+下面我们根据模型设计里的描述定义情感分类模型。其中的`Embedding`实例即嵌入层，在实验中，我们使用了两个嵌入层。`Conv1D`实例即为卷积层，`GlobalMaxPool1D`实例为池化层，卷积层和池化层用于抽取文本中重要的特征。`Dense`实例即生成分类结果的输出层。
+
+
+```
+class TextCNN(nn.Block):
+    def __init__(self, vocab, embedding_size, **kwargs):
+        super(TextCNN, self).__init__(**kwargs)
+        self.FILTERS = [3, 4, 5]
+        self.FILTER_NUM = [100, 100, 100]
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 把filters和filter_num都小写，然后这[3, 4, 5]和[100, 100, 100]几个值通过init的参数从外面传进里面。也就是说，把它们作为超参。

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        self.embedding_static = nn.Embedding(len(vocab), embedding_size)
+        self.embedding_non_static = nn.Embedding(len(vocab), embedding_size)
+        for i in range(len(self.FILTERS)):
+            conv = nn.Conv1D(self.FILTER_NUM[i], kernel_size=self.FILTERS[i], strides=1, activation='relu')
+            pool = nn.GlobalMaxPool1D()
+            setattr(self, f'conv_{i}', conv)
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 解释一下这里相当于self.conv_{i} = conv

File: chapter_natural-language-processing/sentiment-analysis-cnn.md (hunk 7 above)
Comment (astonzhang) [reply chapter_natural-language-processing/sentiment-analysis-cnn.md]: FILTERS -> ngram_filter_sizes

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+            setattr(self, f'pool_{i}', pool)
+        self.dropout = nn.Dropout(0.5)
+        self.decoder = nn.Dense(num_outputs)
+    def forward(self, inputs):
+        embeddings_static = self.embedding_static(inputs).transpose((1,2,0))
+        embeddings_non_static = self.embedding_non_static(inputs).transpose((1,2,0))
+        embeddings = nd.concat(embeddings_static,embeddings_non_static,dim=1)
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 注释补充：这里的dim 1的意义是什么？为什么需要concat

Comment (astonzhang) [reply chapter_natural-language-processing/sentiment-analysis-cnn.md]: 逗号后面需要空格

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+        encoding = [
+            nd.flatten(self.get_pool(i)(self.get_conv(i)(embeddings)))
+            for i in range(len(self.FILTERS))]
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 注释补充：为什么这里用flatten，encoding的形状是什么样的

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+        encoding = nd.concat(*encoding, dim=1)
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 注释：这里的dim1的意义是什么。

**Comment by astonzhang** [comment]
还有，目前的文字部分对于小白还是不太好懂，这样很难看懂图里画的什么意思（比如每个框的形状：为什么每个颜色的框的长和宽是这样的）。实在难以解释的话我考虑一下用数学来一步步解释。

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+batch_size = 10
+embed_size = 100
+    
+net = TextCNN(vocab, embed_size)
+net.initialize(init.Xavier(), ctx=ctx)
+# 设置两个embedding 层的 weight 为预训练的词向量。
+net.embedding_static.weight.set_data(glove_embedding.idx_to_vec.as_in_context(ctx))
+net.embedding_non_static.weight.set_data(glove_embedding.idx_to_vec.as_in_context(ctx))
+# 训练中不更新embedding_static的词向量（net.embedding中的模型参数）。
+net.embedding_non_static.collect_params().setattr('grad_req', 'null')
```
[the remark this answers is not in the document]
Comment (xiaotinghe) [reply chapter_natural-language-processing/sentiment-analysis-cnn.md]: 已修改，谢谢

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+```{.python .input  n=1}
+import sys
+sys.path.append('..')
+import collections
+import gluonbook as gb
+import mxnet as mx
+from mxnet import autograd, gluon, init, metric, nd
+from mxnet.gluon import loss as gloss, nn, rnn
+from mxnet.contrib import text
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: sort alphabetically, contrib is before gluon

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+            setattr(self, f'pool_{i}', pool) #将self.pool_{i}置为第i个pool
+        self.dropout = nn.Dropout(0.5)
+        self.decoder = nn.Dense(num_outputs)
+    def forward(self, inputs):
+        #inputs的维度为(句子长度, batch_size) 
+        
+        #embeddings_static的维度为(句子长度, batch_size, embed_size) 
+        embeddings_static = self.embedding_static(inputs)
+        
+        #Conv1D要求的输入是(batch_size, in_channels, width)。对应到自然语言处理任务中，in_channels即embed_size, width即句子长度。故需要做transpose((1,2,0))
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: Please apply the following changes to all the code comments in the entire file:
1. Leave a space after # symbol. 
2. After the end of the comment, use 。 symbol to mark the end of sentence.
3. English () -> Chinese （）
4. Leave a space before and after English or number. For instance, 你1好Good吗 ->  你 1 好 Good 吗
5. Code comment cannot exceed 80-character-per line limit.

File: chapter_natural-language-processing/sentiment-analysis-cnn.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+        embeddings_static = embeddings_static.transpose((1,2,0))  
+        
+        #embeddings_non_static的维度同上
+        embeddings_non_static = self.embedding_non_static(inputs).transpose((1,2,0))
+        
+        #按照模型设计，每个卷积核都应用于两个channel。拼接这两个嵌入层，dim=1的意思是从in_channels这个维度进行拼接。拼接后的维度是(batch_size, in_channels*2, width)
+        embeddings = nd.concat(embeddings_static, embeddings_non_static, dim=1)
```
Comment (astonzhang) [inline chapter_natural-language-processing/sentiment-analysis-cnn.md]: 为啥要concat static channel和non-static channel？paper的图1中是将这两个channel分开并分别求卷积

**Comment by astonzhang** [comment]
Very good! I will revise the contents before publishing it.
